### DISCUSSION TITLE
안녕하세요 실습 컨테이너 코드 질문 드려요

### DISCUSSION THREAD

**Discussion by sangroklee1118** [yoonjeong-kwon/fastcampus-kubernetes#18] [Q&A] [answered]
안녕하세요 강사님, 쿠버네티스 처음 배우면서 쿠버네티스 입문 강의 너무 잘 수강한것 같습니다.
좋은강의 감사 드립니다.
ch1부터 ch5 실습미션 까지 전부 수강하였는데요, 
전반적인 개념은 잡힌것 같은데 쿠버네티스를 적용해보면서 응용을 할려고 하다보니 어플리케이션을 어떠한 형식으로 구현해야할지 몰라서 실습에서 제공되는 아래 컨테이너 소스와 Dockerfile 을 혹시 공유해 주실수 있을까요 ? 
order에서 payment 서비스 쪽으로 요청하는 부분과 쿠버네티스 ingress로부터 요청을 받는 부분을 어떻게 해야될지 몰라서 여쭤봅니다.
감사합니다. 
```
yoonjeong/my-app:2.0
yoonjeong/home:1.0
yoonjeong/order:1.0
yoonjeong/payment:1.0

yoonjeong/snackbar-delivery:1.0
yoonjeong/snackbar-order:1.0
yoonjeong/snackbar-payment:1.0
yoonjeong/snackbar-order:2.0
```

**Comment by yoonjeong-kwon** [comment, marked as answer]
@sangroklee1118님, 안녕하세요. 반갑습니다 😃 
각 소스코드를 확인할 수 있는 방법을 말씀드릴게요.
`docker exec -it {CONTAINER ID} /bin/sh` 를 이용해서 실행 중인 컨테이너에 접속 후
`ls -al ` 명령어를 치시면 아래와 같은 구조가 보이실거에요.

```bash
$ docker exec -it {CONTAINER ID} /bin/sh
/usr/src/app # ls -al
total 160
drwxr-xr-x    1 root     root          4096 Feb  3  2022 .
drwxr-xr-x    1 root     root          4096 Feb  3  2022 ..
-rw-r--r--    1 root     root            47 Jan 31  2022 .dockerignore
drwxr-xr-x  167 root     root          4096 Feb  3  2022 node_modules
-rw-r--r--    1 root     root        129407 Feb  3  2022 package-lock.json
-rw-r--r--    1 root     root           439 Jan 31  2022 package.json
-rw-r--r--    1 root     root          4079 Feb  3  2022 server.js
drwxr-xr-x    2 root     root          4096 Feb  3  2022 views
```
그다음 `cat server.js`와 같이 출력 명령어를 실행하시면 소스코드를 확인하실 수 있습니다.

실습 프로그램을 만들기 위해 express 라는 Node.js 웹 애플리케이션 프레임워크를 사용했습니다. 
views에는 화면 파일들이 들어있고
server.js에는 클라이언트 요청을 처리하고 각 서비스 API 호출을 위한 코드가 담겨 있습니다.

예를 들어 order에서 payment API를 호출하는 코드는 아래와 같습니다.
order가 요청을 받았을 때 payment로 요청을 전달하여 받은 응답을 order 클라이언트로 반환하는 예제 입니다.
```
app.post('/order/checkout', (req, res) => {
    fetch('http://payment/receipt', {
        method: 'post',
        body: JSON.stringify(req.body),
        headers: {'Content-Type': 'application/json'}
    })
    .then(function(response) {
        return response.json().then(function(json) { return { message: json, url: response.url }})
    })
    .then(function(json) {
        const printMenus = json.message.orderedMenus.map(item => `${item.name} - ${item.price}원`)
        const printReceipt = `
        [영수증]
    ...
```

참고로 이번 강의는 애플리케이션 개발보다는 쿠버네티스 기본에 초점이 맞춰져 있기 때문에
실습 프로그램은 통신을 테스트하고 결과물을 확인할 수 있는 최소한의 수준으로 만들었다는 점을 참고해주세요.^^

아래 Dockerfile은 모든 서비스에서 동일합니다.
```
ARG ARCH=
FROM ${ARCH}node:16.13.2-alpine3.14

ENV NODE_VERSION 16.13.2
RUN mkdir -p /usr/src/app
WORKDIR /usr/src/app
COPY package*.json ./
RUN npm install
RUN apk update
RUN apk --no-cache add curl
RUN apk add iproute2
COPY . .
EXPOSE 8080
CMD ["npm", "start"]
```

이미지를 생성할 때 이 Dockerfile은 위에서 말씀드린 최상위 디렉토리(/usr/src/app)에 두었고
아래와 같은 명령어를 통해 이미지 생성 후 docker repository에 업로드했습니다.
```
docker build -t yoonjeong/snackbar-order:1.0-arm64v8 --build-arg ARCH=arm64v8/ .
docker push yoonjeong/snackbar-order:1.0-arm64v8
```

학습에 도움이 되시길 바라며
또 궁금한 점은 남겨주세요.
감사합니다!

**Reply by sangroklee1118** [reply]
자세한 설명 너무 감사 드립니다. 알려주신 방법대로 해보고 참고하도록 하겠습니다.
